i2c-flash.c

```diff
@@ -36,18 +44,20 @@ ISR(TWI_vect) {
 	}
 
 	// Acknowledge interrupt to let hardware continue.
-	TWCR |= _BV(TWINT);
+	TWCR = _BV(TWEN) | _BV(TWIE) | _BV(TWINT) | _BV(TWEA);
```

Comment:
What's the purpose of this change? I know that, in general, read-modify-write of registers can have unintended side effects, but I don't think that's the case for `TWCR`. The only bit that might potentially cause the hardware to do things when rewritten are `TWSTA` and `TWSTO`, but we don't use either of those. Unless I'm missing something, I'd prefer to keep the old line for clarity.

Reply:
> What's the purpose of this change? I know that, in general, read-modify-write of registers can have unintended side effects, but I don't think that's the case for `TWCR`. The only bits that might potentially cause the hardware to do things when rewritten are `TWSTA` and `TWSTO`, but the only place we use them is in the new case above, which should have an early return anyway. Unless I'm missing something, I'd prefer to keep the old line for clarity.

THIS is what was causing the issue. Read-modify-write is a big no-no here because the read bits are volatile and not meant to be written back.

Reply:
Interesting, I stand corrected. Do you know which specific bit is the issue? Out of the R/W bits in the register, the only ones your version doesn't write are TWSTA and TWSTO. The hardware never sets those according to the datasheet, and they're not write only, so they should read as zero in my version and result in the exact same write you have.

Also, this ISR has to successfully fire many times (once for each byte read) for even a single boot to work. So I'm curious what the error state is and how we enter it.

Reply:
> Interesting, I stand corrected. Do you know which specific bit is the issue?

All three of the other ones I'm writing are needed. I tested.

> Also, this ISR has to successfully fire many times (once for each byte) for even a single boot to work. So I'm curious what the error state is and how we enter it.

When the i2c read wraps around the end of the buffer, the state reported by the bits changes to something other than what we want.

Reply:
> All three of the other ones I'm writing are needed. I tested.

Right, all three of those are for sure needed. I'm asking which extra bits were written in my implementation and broke things. Or, conversely, if any of those three were somehow cleared by hardware, which would also break things. My reading of the datasheet is that, for this register, read-modify-write is equivalent to what you've written. But if this fixed your issue, I must be wrong. I'll do some more investigation on my own and see if I can repro the issue.

Reply:
Let me rephrase: Not explicitly setting any single one of those three breaks it.

Reply:
Huh, so I guess that means that they all get set to zero by hardware in the error state. I'm very curious what it is now. By "wraps around the end of the buffer", do you mean when this check starts hitting the second case and setting `TWDR` to 0?

https://github.com/tchebb/amlogic-hdmiboot-avr/blob/d6a54c207da07e4387b7b763c382becff78c9689/i2c-flash.c#L34

Reply:
I mean when ```offset``` starts being less than ```sizeof``` again after having been greater than or equal, yes.

Reply:
Also, IIRC, not all of them are cleared each time, but sometimes they are. Either way, the library code makes it plain as day that preserving the values of those bits is not the thing to be done here.

Reply:
> For example, the entire reason this repo exists is because the Arduino I2C library can't manage to respond to interrupts fast enough to avoid clock stretching.

I would attribute that to the significantly higher overhead the library has.